File: src/app/searchprovider.rs
// Copyright Sebastian Wiesner <sebastian@swsnr.de>
//
// Licensed under the EUPL
//
// See https://interoperable-europe.ec.europa.eu/collection/eupl/eupl-text-eupl-12

//! Utilities for the search provider of Turn On.

use formatx::formatx;
use glib::{Variant, VariantDict, dpgettext2};
use gnome_app_utils::futures::{FutureExt as _, future};
use gtk::gio::{
    DBusConnection, DBusError, ListStore, Notification, NotificationPriority, RegistrationId,
};
use gtk::prelude::*;

use crate::app::TurnOnApplication;
use crate::config::G_LOG_DOMAIN;
use crate::dbus::searchprovider2::{self, ActivateResult, GetResultMetas, MethodCall};

use super::model::Device;

fn matches_terms<S: AsRef<str>>(device: &Device, terms: &[S]) -> bool {
    let label = device.label().to_lowercase();
    let host = device.host().to_lowercase();
    terms.iter().all(|term| {
        let term = term.as_ref().to_lowercase();
        label.contains(&term) || host.contains(&term)
    })
}

/// Get a result set.
///
/// Return the id of all devices which match all of `terms`, either in their
/// label or their host.
///
/// The ID of a device is simply the stringified position in the list of devices.
fn get_ids_for_terms<S: AsRef<str>>(devices: &ListStore, terms: &[S]) -> Vec<String> {
    devices
        .into_iter()
        .map(|obj| obj.unwrap().downcast::<Device>().unwrap())
        // Enumerate first so that the index is correct
        .enumerate()
        .filter(|(_, device)| matches_terms(device, terms))
        .map(|(i, _)| i.to_string())
        .collect::<Vec<_>>()
}

fn get_result_set<S: AsRef<str>>(app: &TurnOnApplication, terms: &[S]) -> Variant {
    let results = get_ids_for_terms(&app.devices().registered_devices(), terms);
    (results,).into()
}

async fn activate_result(
    app: &TurnOnApplication,
    call: ActivateResult,
) -> Result<Option<Variant>, glib::Error> {
    let device = call
        .identifier
        .parse::<u32>()
        .ok()
        .and_then(|n| app.devices().registered_devices().item(n))
        .map(|o| o.downcast::<Device>().unwrap());
    glib::trace!(
        "Activating device at index {}, device found? {}",
        call.identifier,
        device.is_some()
    );
    match device {
        None => {
            glib::warn!("Failed to find device with id {}", call.identifier);
            Ok(None)
        }
        Some(device) => {
            if let Ok(()) = device.wol().await {
                let notification = Notification::new(&dpgettext2(
                    None,
                    "search-provider.notification.title",
                    "Sent magic packet",
                ));
                notification.set_body(Some(
                    &formatx!(
                        dpgettext2(
                            None,
                            "search-provider.notification.body",
                            "Sent magic packet to mac address {device_mac_address} \
of device {device_label}.",
                        ),
                        device_label = device.label(),
                        device_mac_address = device.mac_address()
                    )
                    .unwrap(),
                ));
                let id = glib::uuid_string_random();
                app.send_notification(Some(&id), &notification);
                glib::timeout_future_seconds(10).await;
                app.withdraw_notification(&id);
            } else {
                let notification = Notification::new(&dpgettext2(
                    None,
                    "search-provider.notification.title",
                    "Failed to send magic packet",
                ));
                notification.set_body(Some(
                    &formatx!(
                        dpgettext2(
                            None,
                            "search-provider.notification.body",
                            "Failed to send magic packet to mac address \
{device_mac_address} of device {device_label}.",
                        ),
                        device_label = device.label(),
                        device_mac_address = device.mac_address()
                    )
                    .unwrap(),
                ));
                notification.set_priority(NotificationPriority::Urgent);
                app.send_notification(None, &notification);
            }
            Ok(None)
        }
    }
}

fn get_result_metas(app: &TurnOnApplication, call: &GetResultMetas) -> Variant {
    let metas: Vec<VariantDict> = call
        .identifiers
        .iter()
        .filter_map(|id| {
            id.parse::<u32>()
                .ok()
                .and_then(|n| app.devices().registered_devices().item(n))
                .map(|obj| {
                    let device = obj.downcast::<Device>().unwrap();
                    let metas = VariantDict::new(None);
                    metas.insert("id", id);
                    metas.insert("name", device.label());
                    metas.insert("description", device.host());
                    metas
                })
        })
        .collect::<Vec<_>>();
    (metas,).into()
}

async fn dispatch_method_call(
    app: TurnOnApplication,
    call: MethodCall,
) -> Result<Option<Variant>, glib::Error> {
    use MethodCall::*;
    match call {
        GetInitialResultSet(c) => {
            glib::trace!("Initial search for terms {:?}", c.terms);
            Ok(Some(get_result_set(&app, c.terms.as_slice())))
        }
        GetSubsearchResultSet(c) => {
            glib::trace!(
                "Sub-search for terms {:?}, with initial results {:?}",
                c.terms,
                c.previous_results
            );
            // We just search fresh again, since our model is neither that big nor that complicated
            Ok(Some(get_result_set(&app, c.terms.as_slice())))
        }
        GetResultMetas(c) => Ok(Some(get_result_metas(&app, &c))),
        ActivateResult(c) => activate_result(&app, c).await,
        LaunchSearch(c) => {
            glib::debug!("Launching search for terms {:?}", &c.terms);
            // We don't have in-app search (yet?) so let's just raise our main window
            app.activate();
            Ok(None)
        }
    }
}

/// Register the Turn On search provider for `app`.
///
/// Register a search provider for devices on the D-Bus connection of `app`.
/// The search provider exposes devices from the `app` model to GNOME Shell,
/// and allows to turn on devices directly from the shell overview.
pub fn register_app_search_provider(
    connection: &DBusConnection,
    app: &TurnOnApplication,
) -> Result<RegistrationId, glib::Error> {
    let interface_info = searchprovider2::interface();
    let registration_id = connection
        .register_object("/de/swsnr/turnon/search", &interface_info)
        .typed_method_call::<searchprovider2::MethodCall>()
        .invoke_and_return_future_local(glib::clone!(
            #[weak_allow_none]
            app,
            move |_, sender, call| {
                glib::debug!("Sender {sender:?} called method {call:?}");
                if let Some(app) = app {
                    dispatch_method_call(app, call).right_future()
                } else {
                    future::ready(Err(glib::Error::new(
                        DBusError::Disconnected,
                        "Application is gone",
                    )))
                    .left_future()
                }
            }
        ))
        .build()?;
    Ok(registration_id)
}

#[cfg(test)]
mod tests {
    use macaddr::MacAddr6;

    use crate::{app::model::Device, net::WOL_DEFAULT_TARGET_ADDRESS};

    use super::*;

    #[test]
    fn device_matches_terms_case_insensitive() {
        let device = Device::new(
            "Server",
            MacAddr6::nil().into(),
            "foo.example.com",
            WOL_DEFAULT_TARGET_ADDRESS.into(),
        );
        assert!(matches_terms(&device, &["server"]));
        assert!(matches_terms(&device, &["SERVER"]));
        assert!(matches_terms(&device, &["SeRvEr"]));
        assert!(matches_terms(&device, &["FOO"]));
        assert!(matches_terms(&device, &["fOo"]));
    }

    #[test]
    fn device_matches_terms_in_label_and_host() {
        let device = Device::new(
            "Server",
            MacAddr6::nil().into(),
            "foo.example.com",
            WOL_DEFAULT_TARGET_ADDRESS.into(),
        );
        assert!(matches_terms(&device, &["Server", "foo"]));
    }

    #[test]
    fn device_matches_terms_ignores_mac_address() {
        let device = Device::new(
            "Server",
            "a2:35:e4:9e:b4:c3".parse().unwrap(),
            "foo.example.com",
            WOL_DEFAULT_TARGET_ADDRESS.into(),
        );
        assert!(!matches_terms(&device, &["a2:35"]));
    }
}

File: src/turnon/searchprovider.py
# Copyright Sebastian Wiesner <sebastian@swsnr.de>
#
# Licensed under the EUPL
#
# See https://interoperable-europe.ec.europa.eu/collection/eupl/eupl-text-eupl-12

"""The search provider for Turn On."""

from dataclasses import asdict, dataclass
from functools import cache
from gettext import pgettext as C_
from typing import cast

from gi.repository import Gio, GLib, GObject

from .dbus import load_dbus_interface
from .model import DeviceObject
from .net import wol


def _check_type(variant: GLib.Variant, expected: str) -> None:
    ts = variant.get_type_string()
    if ts != expected:
        raise GLib.Error(
            domain=GLib.quark_to_string(Gio.DBusError.quark()),
            code=Gio.DBusError.INVALID_ARGS,
            message=f"Expected {expected} but got {ts}",
        )


@dataclass(kw_only=True)
class ResultMeta:
    """Result meta information in the search provider interface."""

    id: str
    name: str
    description: str

    def to_variant(self) -> dict[str, GLib.Variant]:
        """Convert to a variant dict."""
        return {k: GLib.Variant("s", v) for k, v in asdict(self).items()}


@cache
def search_provider_interface() -> Gio.DBusInterfaceInfo:
    """Get the search provider D-Bus interface description."""
    return load_dbus_interface("org.gnome.Shell.SearchProvider2")


def _matches(device: DeviceObject, terms: list[str]) -> bool:
    return all(
        t in device.device.label.lower() or t in device.device.host.lower()
        for t in terms
    )


class SearchProvider(GObject.Object):
    """Search provider for Turn On."""

    __gtype_name__ = "TurnOnSearchProvider"

    def __init__(self, id_prefix: str, devices: Gio.ListModel[DeviceObject]) -> None:
        """Create a search provider for `devices`."""
        super().__init__()
        self._id_prefix = id_prefix
        self._devices = devices

    def _get_initial_result_set(self, terms: list[str]) -> list[str]:
        return [
            f"{self._id_prefix}-{i}"
            for i, device in enumerate(self._devices)
            if _matches(device, terms)
        ]

    def _lookup_by_id(self, id: str) -> DeviceObject | None:
        return self._devices.get_item(int(id[len(self._id_prefix) + 1 :]))

    def _get_subsearch_results(
        self, prev_results: list[str], terms: list[str]
    ) -> list[str]:
        # We don't care for previous results, as our model is small enough
        return [
            id
            for id, device in ((id, self._lookup_by_id(id)) for id in prev_results)
            if device is not None and _matches(device, terms)
        ]

    def _get_result_metas(self, ids: list[str]) -> list[ResultMeta]:
        devices = ((id, self._lookup_by_id(id)) for id in ids)
        return [
            ResultMeta(id=id, name=d.device.label, description=d.device.host)
            for id, d in devices
            if d is not None
        ]

    async def _activate_result(self, id: str) -> None:
        device = self._lookup_by_id(id)
        if device is None:
            raise GLib.Error(
                domain=GLib.quark_to_string(Gio.DBusError.quark()),
                code=Gio.DBusError.INVALID_ARGS,
                message=f"Device with ID {id} not found",
            )
        try:
            await wol(device.device.mac_address, device.device.target_address)
            notification = Gio.Notification.new(
                C_(
                    "search-provider.notification.title",
                    "Sent magic packet",
                )
            )
            notification.set_body(
                C_(
                    "search-provider.notification.body",
                    "Sent magic packet to mac address {device_mac_address} "
                    + "of device {device_label}.",
                ).format(
                    device_mac_address=str(device.device.mac_address),
                    device_label=device.device.label,
                )
            )
            self.emit("send-notification", notification, 10)
        except BaseException:
            notification = Gio.Notification.new(
                C_(
                    "search-provider.notification.title",
                    "Failed to send magic packet",
                )
            )
            notification.set_body(
                C_(
                    "search-provider.notification.body",
                    "Failed to send magic packet to mac address "
                    + "{device_mac_address} of device {device_label}.",
                ).format(
                    device_mac_address=str(device.device.mac_address),
                    device_label=device.device.label,
                )
            )
            self.emit("send-notification", notification, -1)
            raise

    @GObject.Signal(arg_types=[object])  # pyright: ignore[reportUntypedFunctionDecorator]
    def search_launched(self, terms: list[str]) -> None:
        """Signal emitted when search is launched."""
        pass

    @GObject.Signal(arg_types=[Gio.Notification, int])  # pyright: ignore[reportUntypedFunctionDecorator]
    def send_notification(self, notification: Gio.Notification, timeout: int) -> None:
        """Signal emitted when a notification should be shown by the app."""
        pass

    async def call_method(
        self, method: str, params: GLib.Variant
    ) -> GLib.Variant | None:
        """Asynchronously call a method on this search provider."""
        match method:
            case "GetInitialResultSet":
                _check_type(params, "(as)")
                (terms,) = cast(tuple[list[str]], params.unpack())
                ids = self._get_initial_result_set(terms)
                return GLib.Variant("(as)", (ids,))
            case "GetSubsearchResultSet":
                _check_type(params, "(asas)")
                (prev_results, terms) = cast(
                    tuple[list[str], list[str]], params.unpack()
                )
                ids = self._get_subsearch_results(prev_results, terms)
                return GLib.Variant("(as)", (ids,))
            case "GetResultMetas":
                _check_type(params, "(as)")
                (ids,) = cast(tuple[list[str]], params.unpack())
                return GLib.Variant(
                    "(aa{sv})",
                    ([meta.to_variant() for meta in self._get_result_metas(ids)],),
                )
            case "ActivateResult":
                _check_type(params, "(sasu)")
                (id, _, _) = cast(tuple[str, list[str], int], params.unpack())
                await self._activate_result(id)
                return None
            case "LaunchSearch":
                _check_type(params, "(asu)")
                (terms, _) = cast(tuple[list[str], int], params.unpack())
                self.emit("search-launched", terms)
                return
            case other:
                raise GLib.Error(
                    domain=GLib.quark_to_string(Gio.DBusError.quark()),
                    code=Gio.DBusError.UNKNOWN_METHOD,
                    message=f"Unknown method: {other}",
                )
